### PR TITLE
Simplify ComponentModel.forward()

### PR DESCRIPTION
**Pull request by danbraunai**
## Description
- Refactored the ComponentModel.forward() method. Signature now looks like this:
```
    def forward(
        self,
        *args: Any,
        mask_infos: dict[str, ComponentsMaskInfo] | None = None,
        cache_type: Literal["input", "none"] = "none",
        **kwargs: Any,
    ) -> Tensor | CachedOutput:
        """Forward pass with optional component replacement and/or input caching.

        This method handles the following 4 cases:
        1. mask_infos is None and cache_type is "none": Regular forward pass.
        2. mask_infos is None and cache_type is "input": Forward pass with input caching on
            all modules in self.module_paths.
        3. mask_infos is not None and cache_type is "input": Forward pass with component replacement
            and input caching on the modules provided in mask_infos.
        4. mask_infos is not None and cache_type is "none": Forward pass with component replacement
            on the modules provided in mask_infos and no caching.

        We use the same _components_and_cache_hook for cases 2, 3, and 4, and don't use any hooks
        for case 1.

        Args:
            mask_infos: Dictionary mapping module names to ComponentsMaskInfo.
                If provided, those modules will be replaced with their components.
            cache_type: If "input", cache the inputs to the modules provided in mask_infos. If
                mask_infos is None, cache the inputs to all modules in self.module_paths.

        Returns:
            CachedOutput object if cache_type is "input", otherwise the model output tensor.
        """
```
where CachedOutput is a NamedTuple with fields `output` and `cache`.

- Adds overloads for ComponentModel.__call__ so that we get typing on model outputs

Note that we don't bother with `@overload`s on the forward method to handle the different signatures, because we never call "forward" directly.

## Motivation and Context

Partly in preparation for #158, which will require caching the inputs to modules on forward passes with and without component insertion.

Also, our current signature for forward() is messy (we don't need mode="components" when we go through the components iff masks are supplied.

## How Has This Been Tested?
No new tests. Existing runs work.

## Does this PR introduce a breaking change?
No

### CODE REVIEW

**Comment by danbraunai** [comment]
Another option which I think I'm partial to is to just pass a cache object directly as an argument to `__call__`, and have forward() always just output the model. Thoughts @oli-clive-griffin ?

>  You get a good amount of the benefit of the type system by being able to check your correctness by adding .forward temporarily. obviously it's not as robust but it's not useless

As in, while you're developing, call `.forward()` instead of `.__call__()` to verify your correctness? I've never done this before, skeptical that there's a lot of value to this. Also I probably see the downside of @overload as higher than most (I think it's both annoying for engineers and confusing for researchers).

**Review by oli-clive-griffin — APPROVED** [review]
(posted before I saw your comment)

I think this is a good compromise. Did some playing around and couldn't think of anything better.

Re this pattern:
```
pre_weight_acts = model(batch, cache_type="input")[1]
```

could we return a NamedTuple and do:
```
pre_weight_acts = model(batch, cache_type="input").cache
```

instead?

**Comment by oli-clive-griffin** [comment]
> As in, while you're developing, call .forward() instead of .__call__() to verify your correctness? I've never done this before, skeptical that there's a lot of value to this. Also I probably see the downside of @overload as higher than most (I think it's both annoying for engineers and confusing for researchers).

Yea I mean I just temporarily add it in to check typing. However with overloads that starts to get dicey as you're really just relying on the correctness of the overload typing. Wouldn't be against removing them and this just being a part of the code that you need to pay particular attention to when interacting with it

**Comment by danbraunai** [comment]
> could we return a NamedTuple
I think I'd prefer returning a named tuple than having overloads. I think I also prefer it to having different return types without overloads. But it is a little annoying having to do `out = model(batch).output` everywhere.

**Comment by oli-clive-griffin** [comment]
> Another option which I think I'm partial to is to just pass a cache object directly as an argument to __call__, and have forward() always just output the model output and no cache. Thoughts @oli-clive-griffin ?

I'm generally not a fan of this pattern. I think the better version is something like:

```
with compoment_model.input_cache() as cache:
    compoment_model(batch, mask_infos=mask_infos)
```

but after writing that out I don't like it either. I think what we've got is the best compromise, nice and functional and straightforward. And potentially a little nicer if we return a `NamedTuple` as noted above

**Comment by oli-clive-griffin** [comment]
I just mean return a named tuple in the case of caching, and a straight return otherwise

**Comment by oli-clive-griffin** [comment]
(should we maybe just call quickly)

**Comment by danbraunai** [comment]
We discussed on our call how a namedtuple might be nicer in the case where we want some caching. This would mean that we can do `cache = model(batch, cache_type="input").cache` instead of `cache = model(batch, cache_type="input")[1]`.

I agree that this is nicer, however I don't think it's worth the cost of having to do:
```
raw_out = model(batch, cache_type="input")
assert isinstance(raw_out, ModelCacheOutput)
out = raw_out.output
cache = raw_out.cache
```
instead of just `out, cache = model(batch, cache_type="input")`. In most cases we want both the output and the cache.

So I'm inclined to just leave it as an unnamed tuple. Thoughts @oli-clive-griffin ?

**Comment by danbraunai** [comment]
Ready for another review when you're back at work @oli-clive-griffin.

File: spd/metrics/stochastic_recon_layerwise_loss.py [hunk 1]
```diff
@@ -42,10 +42,9 @@ def _stochastic_recon_layerwise_loss_update(
     for stochastic_mask_infos in stochastic_mask_infos_list:
         for module_name, mask_info in stochastic_mask_infos.items():
             out = model(batch, mask_infos={module_name: mask_info})
-            loss_type = output_loss_type
-            loss = calc_sum_recon_loss_lm(pred=out, target=target_out, loss_type=loss_type)
+            loss = calc_sum_recon_loss_lm(pred=out, target=target_out, loss_type=output_loss_type)
 
-            n_examples += out.shape.numel() if loss_type == "mse" else out.shape[:-1].numel()
+            n_examples += out.shape.numel() if output_loss_type == "mse" else out.shape[:-1].numel()
```
[the remark this answers is not in the document]
Comment (danbraunai) [reply spd/metrics/stochastic_recon_layerwise_loss.py:47]: - For KL we don't want to mean over the output dim. We want the "[batchmean](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.kl_div.html)" which means over everything but the output dim. This is just because that's the standard math definition of KL
- For MSE we DO want the mean over the output dim.

Did you have something in mind to tidy this up?
